src/game/logic/object/weaponset.cpp

```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
+        || behavior->Get_Can_Any_Slaves_Use_Weapon_Against_Target(type, victim_obj, location, source) != ATTACK_RESULT_3) {
+        return ATTACK_RESULT_1;
+    }
+
+    if (source_obj->Is_KindOf(KINDOF_IMMOBILE) && source_obj->Is_KindOf(KINDOF_SPAWNS_ARE_THE_WEAPONS)
+        && result == ATTACK_RESULT_2) {
+        return ATTACK_RESULT_3;
+    }
+
+    return result;
+}
+
+bool WeaponSet::Choose_Best_Weapon_For_Target(
+    const Object *source_obj, const Object *victim_obj, WeaponChoiceCriteria criteria, CommandSourceType source)
+{
+    if (Is_Cur_Weapon_Locked()) {
+        return true;
+    }
+
+    if (victim_obj != nullptr) {
+        bool b1 = false;
+        bool b2 = false;
+        float f1 = 0.0f;
+        float f2 = 0.0f;
+        float f3 = 0.0f;
+        float f4 = 0.0f;
+        int i1 = 0;
+        int i2 = 0;
+
+        for (int i = WEAPONSLOT_TERTIARY; i >= 0; i--) {
+            if (m_weapons[i] != nullptr) {
+                int mask = m_curWeaponTemplateSet->Get_Nth_Auto_Choose_Mask(static_cast<WeaponSlotType>(i));
+
+                if (((1 << source) & mask) != 0 || (mask & 4) != 0) {
+                    Weapon *weapon = m_weapons[i];
+
+                    if (weapon != nullptr
+                        && (weapon->Get_Status() != Weapon::OUT_OF_AMMO || weapon->Is_Auto_Reloads_Clip())) {
+                        if ((Get_Victim_Anti_Mask(victim_obj) & weapon->Get_Anti_Mask()) != 0
+                            && weapon->Is_Within_Target_Pitch(source_obj, victim_obj)) {
+                            float damage = weapon->Estimate_Weapon_Damage(source_obj, victim_obj);
+                            float range = weapon->Get_Attack_Range(source_obj);
+                            bool b3 = weapon->Get_Status() == Weapon::READY_TO_FIRE;
+                            const AIUpdateInterface *update = source_obj->Get_AI_Update_Interface();
+
+                            if (update != nullptr
+                                && update->Is_Weapon_Slot_On_Turret_And_Aiming_At_Target(
+                                    static_cast<WeaponSlotType>(i), victim_obj)) {
+                                b3 = false;
+                            }
+
+                            if (damage > 0.0f || weapon->Get_Damage_Type() == DAMAGE_UNRESISTABLE) {
+                                BitFlags<KINDOF_COUNT> against =
+                                    m_curWeaponTemplateSet->Get_Nth_Preferred_Against_Mask(static_cast<WeaponSlotType>(i));
+
+                                if (against.Any() && victim_obj->Is_KindOf_Multi(against, KINDOFMASK_NONE)) {
+                                    damage = 10000000000.0f;
+                                    range = 10000000000.0f;
+                                    b3 = weapon->Get_Status() != Weapon::OUT_OF_AMMO;
+                                }
+
+                                if (criteria != PREFER_MOST_DAMAGE) {
+                                    if (criteria == PREFER_LONGEST_RANGE) {
```

Comment:
`WeaponChoiceCriteria` only has 2 states. Why is there a double test if this has no `else` ?

Reply:
No idea.